### ISSUE TITLE
Preventing zombies from timed out "sudo" checks.

### ISSUE THREAD

**Issue by davideauee** [sensu/sensu#1831]
When running sensu-client as a non privileged user, if a check using sudo times out, it leaves a zombie behind.

## Expected Behavior
no zombies

## Current Behavior
zombies

## Possible Solution
maybe leveraging non-blocking `Process.waitall`.

i.e. in sensu/client/process.rb 
```
...
      def self.run(options={})
        client = self.new(options)
        EM::run do
          client.start
          client.setup_signal_traps
          EM::PeriodicTimer.new(60) { ::Process.waitall }
        end
      end
...
```
(i admit, i haven't tried)

## Steps to Reproduce (for bugs)

* start sensu-client as a non-privileged user
* add the appropriate permissiong in the sudoers file:
  ```
  sensu  ALL=(ALL) NOPASSWD:/bin/sleep
  ```
* setup a check with configuration: 
  ```
  {
    "checks": {
      "zombies": {
        "type": "standard",
        "command": "sudo /bin/sleep 20",
        "timeout": 10,
        "standalone": true,
        "aggregate": false,
        "handle": true,
        "publish": true,
        "interval": 60
      }
    }
  }
  ```
* reload/restart sensu-client, and wait
  ```
  # ps -efH
  ...snip...
  sensu    27059     1  0 11:29 ?        00:00:00   /opt/sensu/embedded/bin/ruby /opt/sensu/bin/sensu-client -b -c /etc/sensu/config.json -d  /etc/sensu/conf.d -e /etc/sensu/extensions -p /var/run/sensu/sensu-client.pid -l /var/log/sensu/sensu-client.log -L info
  root     27246 27059  0 11:30 ?        00:00:00     [sudo] <defunct>
  root     27436 27059  0 11:31 ?        00:00:00     [sudo] <defunct>
  root     27642 27059  0 11:32 ?        00:00:00     sudo /bin/sleep 20
  root     27644 27642  0 11:32 ?        00:00:00       /bin/sleep 20
  ...snip...
  ```

**Comment by portertech** [comment]
When the sensu-client goes to stop (TERM then KILL) the `sudo` process after timeout, it runs into an error `Operation not permitted`. Not sure what our options are. I've started by adding the error to the timeout output message: `"output":"Execution timed out - Unable to TERM/KILL the process: Operation not permitted"`.

**Comment by portertech** [comment]
The sensu-client will never be able to kill the `sudo` process as it is owned by root. The only "solution" to the problem that I am aware of is allowing the sensu-client to `sudo kill` 😭 not a good idea for many reasons (e.g. kill anything and everything) and an implementation would be atrocious.

**Comment by portertech** [comment]
Improved output string PR: https://github.com/sensu/sensu-spawn/pull/25

**Comment by majormoses** [comment]
I agree there is not a whole lot we can do other than improve other than the logging on it I will check out the PR but I'd like to see the actual process ID in the message as this could be something that could be crawled by logs and alerted on and make the responding engineers life just a tag bit easier.

**Comment by portertech** [comment]
Closed by #1849

**Comment by davideauee** [comment]
I know that it is a few months old but I would like to reopen this ticket and get the suggested code in the description re-evaluated.

It is true that sensu will not be able to terminate the process started under sudo but eventually those processes do terminate and a simple wait allow them to get purged rather than having them pilling up.
